scipy/optimize/_trustregion_exact.py

```diff
@@ -0,0 +1,424 @@
+           SIAM Journal on Numerical Analysis, 16(2), 368-375.
+    .. [2] Golub, G. H., Van Loan, C. F. (2013).
+           "Matrix computations". Forth Edition.
+           JHU press. pp. 140-142
+    """
+
+    c = np.atleast_2d(c)
+    m, n = c.shape
+
+    if m != n:
+        raise ValueError("A square triangular matrix should be provided.")
+
+    # A vector `e` with components selected from {+1, -1}
+    # is selected so that the solution `w` to the system
+    # `C.T w = e` is as large as possible. Implementation
+    # based on algorithm 3.5.1, p. 142, from reference [2]
+    # adapted for lower triangular matrix.
+
+    p = np.zeros(n)
+    w = np.empty(n)
+
+    for k in range(n):
+        wp = (1-p[k])/c.T[k, k]
+        wm = (-1-p[k])/c.T[k, k]
+        pp = p[k+1:]+c.T[k+1:, k]*wp
+        pm = p[k+1:]+c.T[k+1:, k]*wm
+
+        if abs(wp)+norm(pp, 1) >= abs(wm)+norm(pm, 1):
+            w[k] = wp
+            p[k+1:] = pp
+        else:
+            w[k] = wm
+            p[k+1:] = pm
+
+    # The system `C v = w` is solved using backward substitution.
+    v = solve_triangular(c, w)
+
+    # Compute vector norms
+    vnorm = norm(v)
+    wnorm = norm(w)
+
+    # Smallest singular value
+    smin = wnorm/vnorm
+
+    # Associated vector
+    zmin = v/vnorm
+
+    return smin, zmin
+
+
+def gershgorin_bounds(H):
+    """
+    Given a square matrix ``M`` compute upper
+    and lower bounds for its eigenvalues (Gregoshgorin Bounds).
+    Defined ref. [1].
+
+    References
+    ----------
+    .. [1] Conn, A. R., Gould, N. I., & Toint, P. L. (2000).
+           "Trust region methods". Siam. pp. 19.
+    """
+
+    D = np.diag(np.diag(H))
+    absHminusD = abs(H-D)
+    ub = max(sum(D+absHminusD))
+    lb = min(sum(D-absHminusD))
+
+    return lb, ub
+
+
+def singular_leading_submatrix(a, c, k):
+    """
+    Compute term that makes the leading ``k`` by ``k``
+    submatrix from `a` singular.
+
+    Parameters
+    ----------
+    a : ndarray
+        Symmetric matrix that is not positive definite.
+    c : ndarray
+        Upper triangular matrix resulting of an incomplete
+        Cholesky decomposition of matrix `a`.
+    k : int
+        Positive integer such that the leading k by k submatrix from
+        `a` is the first non-positive definite leading submatrix.
+
+    Returns
+    -------
+    delta : float
+        Amout that should be added to the element (k, k) of the
+        leading k by k submatrix of ``a`` to make it singular.
+    v : ndarray
+        A vector such that ``v.T B v = 0``. Where B is the matrix A after
+        ``delta`` is added to its element (k, k).
+    """
+
+    # Compute delta
+    delta = np.sum(c[:k-1, k-1]**2) - a[k-1, k-1]
+
+    # Get dimension
+    n = len(a)
+
+    # Inicialize v
+    v = np.zeros(n)
+    v[k-1] = 1
+
+    # Compute the remaining values of v by solving a triangular system
+    if k != 1:
+        v[:k-1] = solve_triangular(c[:k-1, :k-1], -c[:k-1, k-1])
```

Comment:
And about why I stated that ``v.T B v = 0`` instead of ``B v = 0``: 
It is a little detail the book get a little wrong. Using this procedure you 
don't necessarily get  ``B v = 0``, but only ``v.T B v = 0``. When you compute
``B v = 0`` the first ``k`` lines will be actually equal to zero, but the remaining one
may not be. When you multiply this quantity be ``v.T`` you get rid of this non-zero lines
because v have zero elements for ``n-k`` last lines.